pkg/commands/options/publish.go

```diff
@@ -25,6 +25,9 @@ import (
 
 // PublishOptions encapsulates options when publishing.
 type PublishOptions struct {
+	// DockerRepo overrides the KO_DOCKER_REPO environment variable, if present
+	DockerRepo string
```

Comment:
I'd like to push back on this change, or at least consider it in the larger goal of unifying `ko`'s configuration options.

Today we have .ko.yaml, env vars to override .ko.yaml provided by viper, env vars for `KO_DOCKER_REPO`, flags, etc., and I think all in all we don't have a cohesive story for how people should configure the tool.

This one in particular raises alarms because `KO_DOCKER_REPO` is necessary to use `ko` today, and historically we chose to configure it with an env var because we expect each developer to push to their own registry location. Making this overrideable with a flag might encourage teams/tools to document setting this for all calls, which isn't _bad_ necessarily, it's just a departure I want to make sure we're making intentionally.

In the meantime, is it possible to extract this change from this PR?

Reply:
I have similar feelings. I'd like to formalize the differences between project config, user config, and invocation-specific config a bit better, and this goes in the opposite direction (I think).

Reply:
I'd be happy to consider alternatives here, and I agree that formalizing the different config types in ko would be helpful.

Currently, `makePublisher()` grabs the envvar directly, but this feels like depending too much on the execution context in the core implementation logic. I'd expect something that's closer to the command-line logic to have parsed the config option at this point.

Do you have any alternative recommendations on how to allow the repo to be set programmatically?

For some context, Skaffold has a similar 'default repo' concept, which can be set as a flag, envvar, or in global config: https://skaffold.dev/docs/environment/image-registries/

In the meantime, I'll extract a separate PR for this change.

Reply:
New PR for programmatically setting the destination image repo: https://github.com/google/ko/pull/351